proposals/0000-overloaded-record-fields.rst

```diff
@@ -0,0 +1,359 @@
+This works by giving the occurrence of ``#personId`` the type
+``r -> a`` and introducing a constraint ``HasField "personId" r a``,
+meaning that ``r`` must be some type with a field ``personId :: a``.
+Unification will determine that ``r ~ Person``, and then ``HasField
+"personId" Person a`` can be solved automatically, setting ``a ~ Int``
+in the process.
+
+In addition to resolving ambiguous field selectors to a single record
+type using type inference, the proposed change enables definitions
+that are polymorphic over record fields, so for example
+
+::
+
+  \ p -> #personId p + (1 :: Int)
+
+will be given the inferred type
+
+::
+
+  HasField "personId" r Int => r -> Int
+
+and can be applied to any record type with a field ``personId ::
+Int``.
+
+
+Proposed Change
+---------------
+
+HasField class and OverloadedRecordFields extension
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The module ``GHC.Records`` defines the following class:
+
+::
+
+  class HasField (x :: Symbol) r a | x r -> a where
+    fromLabel :: r -> a
+
+When the new extension ``OverloadedRecordFields`` is enabled:
+
+* the existing ``DuplicateRecordFields`` extension is automatically enabled,
+* overloaded label syntax is available, and
+* a label ``#foo`` is translated to
+  ``GHC.Records.fromLabel @"foo" :: HasField "foo" r a => r -> a``.
+
+The third point assumes ``OverloadedLabels`` and ``RebindableSyntax``
+are *disabled* (see `Multiple versions of fromLabel`_ below).
+
+
+Solving HasField constraints
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+If the constraint solver encounters a constraint ``HasField x r a``
+where ``r`` is a concrete datatype with a field ``x`` in scope, it
+will automatically solve the constraint using the field selector as
+the dictionary, unifying ``a`` with the type of the field if
+necessary.  This happens irrespective of which extensions are enabled,
+as with other built-in classes with special constraint solving
+behaviour (e.g. ``Coercible``).
+
+For the ``personId`` example above, the end result is rather like
+having an instance
+
+::
+
+  instance HasField "personId" Person Int where
+    fromLabel = personId
+
+except that this instance is not actually generated anywhere, rather
+the constraint is solved directly by the constraint solver.
+
+A field must be in scope for the corresponding ``HasField`` constraint
+to be solved.  This retains the existing representation hiding
+mechanism, whereby a module may choose not to export a field,
+preventing client modules from accessing or updating it directly.
+Thus we cannot actually generate and export ``HasField`` instances
+from defining modules: since there is no mechanism for limiting the
+scope of instances, doing so would expose representation details that
+should be hidden.
+
+Users may define their own instances of ``HasField``, subject to the
+usual rules about overlapping and incoherent instances.  This allows
+"virtual" record fields to be defined for datatypes that do not
```

Comment:
I suppose you could change it to:

```haskell
class SetFields (ls :: [Symbol]) s t bs | s ls bs -> t, t ls -> bs where
    setFields :: s -> bs -> t
```

where `bs` is a tuple with `length ls` elements.

```haskell
f t = t { x = True, y = True }
```

will desugar to:

```haskell
f t :: SetFields '["a", "b"] (T a) (T bool) (Bool, Bool) => T a -> T Bool
f t = setFields @'["a", "b"] t (True, True)
```
There are probably a lot of details that need to be sorted out with this approach: for starters, `bs` really should be a "demoted" type-level list of `Type`, because 1-tuple overlaps with other types easily.

For reference, this is very similar what my [`rawr`](https://github.com/PkmX/rawr) library does (`:<=` is the record update operator):

```haskell
type T a = R ( "x" := a, "y" := a )

f :: T a -> T Bool
f t = t :<= R ( #x := True, #y := True )

g :: T a -> T Bool
g t = t :<= R ( #x := True ) -- type error
```